nexus/db-model/src/vpc_firewall_rule.rs

```diff
@@ -54,19 +54,41 @@ impl_enum_wrapper!(
 NewtypeFrom! { () pub struct VpcFirewallRuleAction(external::VpcFirewallRuleAction); }
 NewtypeDeref! { () pub struct VpcFirewallRuleAction(external::VpcFirewallRuleAction); }
 
-impl_enum_wrapper!(
-    VpcFirewallRuleProtocolEnum:
-
-    #[derive(Clone, Debug, AsExpression, FromSqlRow, Serialize, Deserialize)]
-    pub struct VpcFirewallRuleProtocol(pub external::VpcFirewallRuleProtocol);
-
-    Tcp => b"TCP"
-    Udp => b"UDP"
-    Icmp => b"ICMP"
-);
+/// Newtype wrapper around [`external::VpcFirewallRuleProtocol`] so we can derive
+/// diesel traits for it
+#[derive(Clone, Debug, AsExpression, FromSqlRow, Serialize, Deserialize)]
+#[diesel(sql_type = sql_types::Text)]
+#[repr(transparent)]
+pub struct VpcFirewallRuleProtocol(pub external::VpcFirewallRuleProtocol);
 NewtypeFrom! { () pub struct VpcFirewallRuleProtocol(external::VpcFirewallRuleProtocol); }
 NewtypeDeref! { () pub struct VpcFirewallRuleProtocol(external::VpcFirewallRuleProtocol); }
 
+impl ToSql<sql_types::Text, Pg> for VpcFirewallRuleProtocol {
+    fn to_sql<'a>(
+        &'a self,
+        out: &mut serialize::Output<'a, '_, Pg>,
+    ) -> serialize::Result {
+        <String as ToSql<sql_types::Text, Pg>>::to_sql(
+            &self.0.to_string(),
+            &mut out.reborrow(),
+        )
+    }
+}
+
+// Deserialize the "VpcFirewallRuleProtocol" object from SQL TEXT.
+impl<DB> FromSql<sql_types::Text, DB> for VpcFirewallRuleProtocol
+where
+    DB: Backend,
+    String: FromSql<sql_types::Text, DB>,
+{
+    fn from_sql(bytes: DB::RawValue<'_>) -> deserialize::Result<Self> {
+        Ok(VpcFirewallRuleProtocol(
+            String::from_sql(bytes)?
+                .parse::<external::VpcFirewallRuleProtocol>()?,
```

Comment:
i had a vague memory of there being some way to do this by turning the bytes into an `&str` with `str::from_utf8` or something, so that we don't make a whole `String` just to parse it and throw it away...but i wasn't able to find prior art for that elsewhere in Nexus, so i dunno if it's worth the effort.

Reply:
I put a little time into this while waiting on some racklette-related things: https://github.com/oxidecomputer/omicron/pull/8303